analyzers/src/SonarAnalyzer.Common/Helpers/GlobPatternMatcher.cs

```diff
@@ -0,0 +1,148 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Concurrent;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace SonarAnalyzer.Helpers;
+
+public interface IGlobPatternMatcher
+{
+    bool IsMatch(string pattern, string input);
+}
+
+public class GlobPatternMatcher : IGlobPatternMatcher
+{
+    public bool IsMatch(string pattern, string input) =>
+        WildcardPattern.Create(pattern).Match(input);
+
+    /// <summary>
+    /// Copied from https://github.com/SonarSource/sonar-plugin-api/blob/a9bd7ff48f0f77811ed909070030678c443c975a/sonar-plugin-api/src/main/java/org/sonar/api/utils/WildcardPattern.java
+    /// </summary>
+    private sealed class WildcardPattern
+    {
+        private const string SPECIALCHARS = "()[]^$.{}+|";
+        private static readonly ConcurrentDictionary<string, WildcardPattern> CACHE = new();
+        internal readonly Regex Pattern;
+
+        private WildcardPattern(string pattern, string directorySeparator) =>
+            Pattern = new Regex(ToRegexp(pattern, directorySeparator), RegexOptions.None, TimeSpan.FromSeconds(10));
```

Comment:
I suggest having a more strict timeout like; FromMilliseconds(100). Also keep in mind that if the timeout happens it will throw an exception. So, we need to handle it, not to have AD0001 issues.

Reply:
Error handling for the RegexMatchTimeout is still missing. You can take an inspiration from here: https://github.com/SonarSource/sonar-text/pull/107/files